File: MemoryCompareResult.cpp
#include "MemoryCompareResult.hpp"
#include "MemoryCompare.hpp"
#include <filesystem>

std::wstring MemoryCompare::MemCompareResults::_path = L"";

MemoryCompare::MemCompareResults::MemCompareResults(std::wstring path, uint16_t iteration, uint8_t addressWidth, uint16_t valueWidth, uint16_t rangeCount, bool cached, bool zip)
{
	_path = path;
	_rangeCount = rangeCount;
	_iteration = iteration;
	_addressWidth = addressWidth;
	_valueWidth = valueWidth;
	_cached = cached;
	_zipped = zip;
}

//MemoryCompare::MemCompareResults::MemCompareResults(const MemCompareResults& other)
//{
//	*this = other;
//}

void MemoryCompare::MemCompareResults::createHeader(const uint32_t index)
{
	_fileHeaders.emplace_back();
	_fileHeaders[index].resize(16);
	*(uint64_t*)_fileHeaders[index].data() = _startingAdresses[index];
	*(uint64_t*)(_fileHeaders[index].data() + RESULT_COUNT) = _resultCounts[index];
	*(uint16_t*)(_fileHeaders[index].data() + ITERATION) = _iteration;
	*(uint16_t*)(_fileHeaders[index].data() + VALUE_WIDTH) = _valueWidth;
	*(char*)(_fileHeaders[index].data() + ADDRESS_WIDTH) = _addressWidth;
	*(uint64_t*)(_fileHeaders[index].data() + PLACEHOLDER) = 0;
}

void MemoryCompare::MemCompareResults::SetNewRange(const uint64_t startingAddress)
{
	++_rangeCount;
	_startingAdresses.push_back(startingAddress);
	_resultCounts.push_back(0);
	_indices.push_back(_totalResultCount);
}

int MemoryCompare::MemCompareResults::GetRangeIndexOfStartingAddress(const uint64_t startingAddress) const
{
	for (int i = 0; i < _rangeCount; ++i)
		if (_startingAdresses[i] == startingAddress)
			return i;

	return -1;
}

uint64_t MemoryCompare::MemCompareResults::GetTotalResultCount()
{
	return _totalResultCount;
}

uint64_t MemoryCompare::MemCompareResults::GetResultCountByRangeIndex(const uint32_t index)
{
	return _resultCounts[index];
}

uint64_t MemoryCompare::MemCompareResults::GetResultCountOfRange(const uint64_t startingAddress)
{
	for (int i = 0; i < _rangeCount; ++i)
		if (_startingAdresses[i] == startingAddress)
			return _resultCounts[i];

	return _resultCounts[0];
}

void MemoryCompare::MemCompareResults::Clear(const bool keepIterationCount)
{
	_fileHeaders.clear();
	_addresses.clear();
	_values.clear();
	_previousValues.clear();
	_resultCounts.clear();
	_startingAdresses.clear();
	_totalResultCount = 0;

	if(!keepIterationCount)
		_iteration = 0;
}

void MemoryCompare::MemCompareResults::SetValueWidth(const uint16_t width)
{
	_valueWidth = width;
}

void MemoryCompare::MemCompareResults::SetAddressWidth(const uint8_t width)
{
	_addressWidth = width;
}

bool MemoryCompare::MemCompareResults::SaveResults(uint32_t rangeIndex, bool zipped)
{
	if (rangeIndex == -1)
		return false;

	createHeader(rangeIndex);
	std::wstring path = _path + std::to_wstring(_iteration-1);
	std::filesystem::create_directories(path);
	path.append(L"\\" + std::to_wstring(rangeIndex) + L".bin");

	if (!SaveBinary(path, _fileHeaders[rangeIndex].data(), HEADER_SIZE, false, zipped))
		return false;

	if (_addresses.size() == 0)
		return true;

	if (!SaveBinary(path, _addresses.data(), _resultCounts[rangeIndex] * _addressWidth, true, zipped))
		return false;

	if (!SaveBinary(path, _values.data(), _resultCounts[rangeIndex] * _valueWidth, true, zipped))
		return false;

	if(_iteration > 1)
		if (!SaveBinary(path, _previousValues.data(), _resultCounts[rangeIndex] * _valueWidth, true, zipped))
			return false;

	return true;
}

void MemoryCompare::MemCompareResults::ClearResultsDir(const int iterationIndex, const bool removeIterationDir)
{
	if (iterationIndex < 0)
	{
		for (const auto& entry : std::filesystem::directory_iterator(_path))
			std::filesystem::remove_all(entry.path());
	}
	else 
	{
		for (const auto& entry : std::filesystem::directory_iterator(_path + std::to_wstring(iterationIndex)))
			std::filesystem::remove_all(entry.path());

		if (removeIterationDir)
		{
			std::filesystem::remove(_path + std::to_wstring(iterationIndex));
		}
	}
}

void MemoryCompare::MemCompareResults::SetResultsDir(const std::wstring& dir)
{
	_path = dir;
}

bool MemoryCompare::MemCompareResults::LoadResults(bool zipped)
{
	Clear(true);
	std::wstring path = _path + std::to_wstring(_iteration-1);

	for (const auto& entry : std::filesystem::directory_iterator(path))
	{
		//std::cout << entry.path() << std::endl;

		uint64_t readSize = HEADER_SIZE;
		_fileHeaders.emplace_back(readSize, 0);
		static int index = _fileHeaders.size() - 1;
		static uint64_t readPos = 0;

		if (!LoadBinary(entry.path(), _fileHeaders.back().data(), readSize, readPos))
			return false;

		_startingAdresses.push_back(*(uint64_t*)_fileHeaders[index].data());
		_resultCounts.push_back(*(uint64_t*)(_fileHeaders[index].data() + RESULT_COUNT));
		_totalResultCount += _resultCounts.back();

		if (index == 0)
		{
			_addressWidth = *(_fileHeaders[index].data() + ADDRESS_WIDTH);
			_valueWidth = *(_fileHeaders[index].data() + VALUE_WIDTH);
			_iteration = *(uint16_t*)(_fileHeaders[index].data() + ITERATION);
		}

		_addresses.emplace_back();
		readSize = _addressWidth * _resultCounts.back();
		_addresses.resize(_addressWidth * _resultCounts.back());
		readPos = HEADER_SIZE;

		if (!LoadBinary(entry.path(), _addresses.data(), readSize, readPos))
			return false;


		_values.emplace_back();
		readPos += readSize;
		readSize = _valueWidth * _resultCounts.back();
		_values.resize(_valueWidth * _resultCounts.back());

		if (!LoadBinary(entry.path(), _values.data(), readSize, readPos))
			return false;

		if (_iteration)
		{
			_previousValues.emplace_back();
			readPos += readSize;
			_previousValues.resize(_valueWidth * _resultCounts.back());

			if (!LoadBinary(entry.path(), _previousValues.data(), readSize, readPos))
				return false;
		}
	}

	return true;
}

uint16_t MemoryCompare::MemCompareResults::GetValueWidth() const
{
	return _valueWidth;
}

uint16_t MemoryCompare::MemCompareResults::GetRangeCount() const
{
	return _rangeCount;
}

void MemoryCompare::MemCompareResults::operator=(const MemCompareResults& other)
{
	_addresses = other._addresses;
	_values = other._values;
	_previousValues = other._previousValues;
	_startingAdresses = other._startingAdresses;
	_resultCounts = other._resultCounts;
	_fileHeaders = other._fileHeaders;
	_path = other._path;
	_rangeCount = other._rangeCount;
	_totalResultCount = other._totalResultCount;
	_addressWidth = other._addressWidth;
	_valueWidth = other._valueWidth;
	_fileSize = other._fileSize;
	_iteration = other._iteration;
	_cached = other._cached;
	_zipped = other._zipped;
}

File: MemoryCompareResult.py
import os, shutil
from typing import List, Optional

class MemCompareResult:
    DUMP_STARTING_ADDRESS = 0
    RESULT_COUNT = 8
    ITERATION = 0x10
    VALUE_WIDTH = 0x12
    ADDRESS_WIDTH = 0x14
    PLACEHOLDER = 0x15
    HEADER_SIZE = 0x20

    _path: str = ""

    def __init__(self, path: str, iteration: int, address_width: int, value_width: int, range_count: int = 0, cached: bool = False, zip: bool = False):
        self._addresses: List[bytes] = []
        self._values: List[bytes] = []
        self._previous_values: List[bytes] = []
        self._indices: List[int] = []
        self._starting_addresses: List[int] = []
        self._result_counts: List[int] = []
        self._file_headers: List[str] = []
        self._path = path
        self._range_count = range_count
        self._total_result_count = 0
        self._address_width = address_width
        self._value_width = value_width
        self._file_size = 0
        self._iteration = iteration
        self._cached = cached
        self._zipped = zip

    def create_header(self, index: int):
        header = bytearray(16)
        header[0:8] = self._starting_addresses[index].to_bytes(8, 'little')
        header[8:16] = self._result_counts[index].to_bytes(8, 'little')
        header[16:18] = self._iteration.to_bytes(2, 'little')
        header[18:20] = self._value_width.to_bytes(2, 'little')
        header[20:21] = self._address_width.to_bytes(1, 'little')
        header[21:29] = (0).to_bytes(8, 'little')
        self._file_headers.append(header)

    def set_new_range(self, starting_address: int):
        self._starting_addresses.append(starting_address)
        self._result_counts.append(0)
        self._indices.append(len(self._addresses) // self._address_width)

    def push_back_result_by_ptr(self, address, value: bytes, previous_value: Optional[bytes] = None):
        self._addresses.extend(address.to_bytes(self._address_width, 'little'))
        self._values.extend(value[:self._value_width])

        if self._iteration > 1 and previous_value is not None:
            self._previous_values.extend(previous_value[:self._value_width])

        self._result_counts[-1] += 1
        self._total_result_count += 1

    def get_range_index_of_starting_address(self, starting_address: int) -> int:
        try:
            return self._starting_addresses.index(starting_address)
        except ValueError:
            return -1

    def get_addresses_ptr_all_ranges(self):
        return self._addresses

    def get_values_ptr_all_ranges(self):
        return self._values

    def get_previous_values_ptr_all_ranges(self):
        return self._previous_values

    def get_addresses_ptr_of_range(self, starting_address: int):
        index = self.get_range_index_of_starting_address(starting_address)
        if index == -1 or self._result_counts[index] == 0:
            return None
        return self._addresses[self._indices[index] * self._address_width:]

    def get_values_ptr_of_range(self, starting_address: int):
        index = self.get_range_index_of_starting_address(starting_address)
        if index == -1 or self._result_counts[index] == 0:
            return None
        return self._values[self._indices[index] * self._value_width:]

    def get_previous_values_ptr_of_range(self, starting_address: int):
        index = self.get_range_index_of_starting_address(starting_address)
        if index == -1 or self._result_counts[index] == 0:
            return None
        return self._previous_values[self._indices[index] * self._value_width:]

    def get_total_result_count(self) -> int:
        return self._total_result_count

    def get_result_count_by_range_index(self, index: int) -> int:
        return self._result_counts[index]

    def get_result_count_of_range(self, starting_address: int) -> int:
        index = self.get_range_index_of_starting_address(starting_address)
        if index == -1:
            return 0
        return self._result_counts[index]

    def clear(self, keep_iteration_count: bool = False):
        self._addresses.clear()
        self._values.clear()
        self._previous_values.clear()
        self._indices.clear()
        self._starting_addresses.clear()
        self._result_counts.clear()
        self._file_headers.clear()
        self._total_result_count = 0
        if not keep_iteration_count:
            self._iteration = 0

    def set_value_width(self, width: int):
        self._value_width = width

    def set_address_width(self, width: int):
        self._address_width = width

    def save_results(self, range_index: int, zipped: bool) -> bool:
        if range_index == -1:
            return False

        self.create_header(range_index)
        path = os.path.join(self._path, str(self._iteration - 1))
        os.makedirs(path, exist_ok=True)
        path = os.path.join(path, f"{range_index}.bin")

        if not self.save_binary(path, self._file_headers[range_index], self.HEADER_SIZE, False, zipped):
            return False

        if not self._addresses:
            return True

        if not self.save_binary(path, self._addresses, self._result_counts[range_index] * self._address_width, True, zipped):
            return False

        if not self.save_binary(path, self._values, self._result_counts[range_index] * self._value_width, True, zipped):
            return False

        if self._iteration > 1:
            if not self.save_binary(path, self._previous_values, self._result_counts[range_index] * self._value_width, True, zipped):
                return False

        return True

    @staticmethod
    def save_binary(file_path: str, data: bytes, size: int, append: bool, zip: bool) -> bool:
        mode = 'ab' if append else 'wb'
        try:
            with open(file_path, mode) as file:
                file.write(data[:size])
            return True
        except IOError:
            return False

    @staticmethod
    def load_binary(file_path: str, out: bytearray, size: int, start_pos: int) -> bool:
        try:
            with open(file_path, 'rb') as file:
                file.seek(start_pos)
                out.extend(file.read(size))
            return True
        except IOError:
            return False

    def get_value_width(self) -> int:
        return self._value_width

    def get_range_count(self) -> int:
        return self._range_count

    @staticmethod
    def clear_results_dir(iteration_index: int, remove_iteration_dir: bool = False):
        # Construct the path to the iteration directory
        iteration_dir = os.path.join(MemCompareResults._path, str(iteration_index))
        
        # Check if the directory exists
        if os.path.exists(iteration_dir):
            if remove_iteration_dir:
                # Remove the entire directory
                shutil.rmtree(iteration_dir)
            else:
                # Remove only the files within the directory
                for filename in os.listdir(iteration_dir):
                    file_path = os.path.join(iteration_dir, filename)
                    try:
                        if os.path.isfile(file_path) or os.path.islink(file_path):
                            os.unlink(file_path)
                        elif os.path.isdir(file_path):
                            shutil.rmtree(file_path)
                    except Exception as e:
                        print(f'Failed to delete {file_path}. Reason: {e}')

    @staticmethod
    def set_results_dir(dir: str):
        MemCompareResults._path = dir